DDF-2125 Add the ability to inject default values into metacards during ingest

jrnorth opened the pull request:
#### What does this PR do?
Add the ability to inject default values into metacards during ingest.
- Move the validation JSON parser into its own bundle
- Update the validation JSON parser to handle default attribute values
- Add a default attribute value API
- Updates the catalog framework to inject default values for missing attributes on creates and updates

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@rzwiefel @glenhein 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith @kcwire

#### How should this be tested?
Create a JSON file that includes some default attribute values as shown in the documentation, ingest some metacards, and verify that the default values were added where appropriate.

#### What are the relevant tickets?
[DDF-2125](https://codice.atlassian.net/browse/DDF-2125)

#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

glenhein commented on catalog/core/catalog-core-defaultvalues/src/main/java/ddf/catalog/data/defaultvalues/DefaultAttributeValueRegistryImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    public Optional<Set<String>> getAttributesWithDefaults(String metacardTypeName) {
+        Validate.notNull(metacardTypeName, "The metacard type name cannot be null.");
```
Should Validate.notNull be a static import?

glenhein commented on catalog/core/catalog-core-defaultvalues/src/test/java/ddf/catalog/data/defaultvalues/DefaultAttributeValueRegistryImplTest.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Test
+    public void testSetDefaultValue() {
+        registry.setDefaultValue(BASIC_METACARD_NAME, TITLE, "foo");
```
You use the "foo", "bar" and "foobar" strings a lot. How about making them static final? And compute the concat version from the first two.

glenhein commented on catalog/plugin/catalog-plugin-defaultvalues/src/main/java/ddf/catalog/plugin/defaultvalues/DefaultValuesPlugin.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        return true;
+    }
```
I prefer the shorter:

return attribute != null ? attribute.getValue() == null : true;

jlcsmith commented on catalog/catalog-app/src/main/resources/features.xml:
```diff
@@ -345,6 +345,20 @@
         <feature>catalog-plugin-metacard-validation</feature>
     </feature>
 
+    <feature name="catalog-core-validationparser" install="manual" version="${project.version}"
```
Any reason we wouldn't want this (and the core-validator feature) installed by default?

bdeining commented on catalog/plugin/catalog-plugin-defaultvalues/src/main/java/ddf/catalog/plugin/defaultvalues/DefaultValuesPlugin.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private void setDefaultValues(Metacard metacard) {
+        // TODO (jrnorth) - What if the attribute is there but has a null value?
```
Does `hasNoValue(Attribute attribute)` address this TODO?

jrnorth replied:
It does.

coyotesqrl commented on catalog/core/catalog-core-api/src/main/java/ddf/catalog/data/DefaultAttributeValueRegistry.java:
```diff
@@ -0,0 +1,95 @@
+     * @param defaultValue  the default value, cannot be null
+     * @throws IllegalArgumentException if either argument is null
+     */
+    void setDefaultValue(String attributeName, Serializable defaultValue);
```
From a contractual point of view, what happens if I call:
1. `setDefaultValue("foobar", "baz");`
2. `setDefaultValue("myspecialmetatype", "foobar", "special snowflake");

How does that affect the semantics of the get and remove methods?

coyotesqrl commented:
This note in the documentation makes the intent clear between global and specific default values:
> In the event that an attribute has a 'global' default value as well as a default value for a specific metacard type, the default value for the specific metacard type will be applied (i.e., the more specific default value wins).

However, the `DefaultAttributeValueRegistry` interface does not clarify that point and its implementation class does not enforce that. Rather, that's enforced by the `CatalogFrameworkImpl` and that feels both a bit brittle and like leakage to me.

jrnorth commented:
@coyotesqrl Yep. Updated the registry API to have only one retrieval method and the implementation to return the proper (global or specific) default value. Also added more javadoc to the interface.

coyotesqrl commented:
👍

bdeining commented:
:+1:

jlcsmith commented:
Built / tested the update logic and works as expected.

jlcsmith commented:
@jrnorth - please rebase with master to resolve conflicts.

jlcsmith commented:
Built and tested.  Looks good to me.

rzwiefel commented:
:+1:

kcwire commented:
👍
